In [1]:
import pandas as pd

In [2]:
titanic_df = pd.read_csv('train.csv')
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [3]:
titanic_df.shape

(891, 12)

In [4]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


<class 'pandas.core.frame.DataFrame'>

RangeIndex: 891 entries, 0 to 890  

Data columns (total 12 columns):

    RangeIndex는 DataFrame index의 범위를 나타내므로 전체 row수를 알 수 있다.

    전체 데이터는 891개 row이다. 그리고 Column 수는 12개

PassengerId    891 non-null int64

    컬럼별 데이터 타입을 알 수 있다. object는 문자열이라고 생각해도 무방

Survived       891 non-null int64

Pclass         891 non-null int64

Name           891 non-null object

Sex            891 non-null object

Age            714 non-null float64

    Age가 714라는 값은 891개중 714개는 값이 있고 나머지 177개는 Null이라는 것을 의미

SibSp          891 non-null int64

Parch          891 non-null int64

Ticket         891 non-null object

Fare           891 non-null float64

Cabin          204 non-null object

Embarked       889 non-null object

dtypes: float64(2), int64(5), object(5)

    전체 12개의 칼럼을 요약

memory usage: 83.6+ KB


In [5]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
# 데이터의 분포도 확인
value_counts = titanic_df['Pclass'].value_counts()
value_counts

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [7]:
titanic_pclass = titanic_df['Pclass']
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [8]:
import numpy as np

col_name1 = ['col1']
list1 = [1,2,3]
array1 = np.array(list1)
print('arrayl shape:', array1.shape)

# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns = col_name1)
print('1차원 리스트로 만든 DataFrame: \n', df_list1)

# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns = col_name1)
print('1차원 ndarry로 만든 DataFrame: \n', df_array1)

arrayl shape: (3,)
1차원 리스트로 만든 DataFrame: 
    col1
0     1
1     2
2     3
1차원 ndarry로 만든 DataFrame: 
    col1
0     1
1     2
2     3


In [9]:
# 3개의 컬럼명이 필요함
col_name2 = ['col1', 'col2', 'col3']

# 2행 * 3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2 = [[1,2,3],
        [11,12,13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)

df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame: \n',df_list2)

df_array2 = pd.DataFrame(array2, columns = col_name2)
print('2차원 ndarray로 만든 DataFrame: \n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13


In [10]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {'col1' : [1,11], 'col2' : [2,22], 'col3' : [3,33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame: \n', df_dict)

# DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입: ', type(array3), '\ndf_dict.values shape: ', array3.shape)
print(array3)

# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입 :', type(list3))
print(list3)

# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입: ', type(dict3))
print(dict3)

딕셔너리로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    22    33
df_dict.values 타입:  <class 'numpy.ndarray'> 
df_dict.values shape:  (2, 3)
[[ 1  2  3]
 [11 22 33]]
df_dict.values.tolist() 타입 : <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입:  <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


In [11]:
titanic_df['Age_0'] = 0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [12]:
titanic_df['Age_by_10'] = titanic_df['Age'] * 10
titanic_df['Family_N0'] = titanic_df['SibSp'] + titanic_df['Parch'] + 1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_N0  
0          2  
1          2  
2          1

In [13]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_N0  
0          2  
1          2  
2          1

#### DataFrame 데이터 삭제

DataFrame.drop(labels=None, axis=0, index=Nonde, columns=None, level=None, inplace=False, errors='raise')

In [14]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_N0  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

In [15]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_N0'], axis=1, inplace=True) # inplace=True => titanic_df내의 데이터도 삭제
print('inplace = True 로 drop후 반환된 값 :', drop_result)
titanic_df.head(3)

inplace = True 로 drop후 반환된 값 : None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [16]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0,1,2], axis=0, inplace=True)
print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


In [17]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('train.csv')

# Index 객체 추출
indexes = titanic_df.index
print(indexes)

# Index 객체를 실제 값 array로 변환
print('Index 객체 array값: \n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값: 
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 23

In [18]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


In [20]:
# Index이므로 값 변경 불가능
# indexes[0] = 5

In [21]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n',(series_fair + 3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


In [22]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [26]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts() # 각 value값 counts
print(value_counts)
print('value_counts 객체 변수 타입: ', type(value_counts))

print()

new_value_counts = value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입:  <class 'pandas.core.series.Series'>

### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


In [29]:
print('단일 칼럼 데이터 추출 :\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[['Survived', 'Pclass']].head(3))

print()
print('[] 안에 숫자 index는 KeyError 오류 발생:\n') # titanic_df[0]) => 오류 발생

단일 칼럼 데이터 추출 :
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3

[] 안에 숫자 index는 KeyError 오류 발생:



In [34]:
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

In [35]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

In [36]:
data = {'Name' : ['Chulmin', 'Eunkyung', 'Jinwoong', 'Soobeom'],
       'Year' : [2011,2016,2015,2015],
       'Gender' : ['male', 'Female', 'Male', 'Male']
       }
data_df = pd.DataFrame(data,index=['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [37]:
# data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset = data_df.reset_index()
data_df_reset = data_df_reset.rename(columns={'index' : 'old_index'})

# 인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스값 생성
data_df_reset.index = data_df_reset.index + 1
data_df_reset

old_index      Name  Year  Gender
1       one   Chulmin  2011    male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

In [39]:
titanic_df = pd.read_csv('train.csv')
titanic_boolean = titanic_df[titanic_df['Age'] > 60]
print(type(titanic_boolean))
titanic_boolean.head()

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch      Ticket     Fare Cabin Embarked
33            34         0       2  Wheadon, Mr...  male  66.0      0      0  C.A. 24579  10.5000   NaN        S
54            55         0       1  Ostby, Mr. ...  male  65.0      0      1      113509  61.9792   B30        C
96            97         0       1  Goldschmidt...  male  71.0      0      0    PC 17754  34.6542    A5        C
116          117         0       3  Connors, Mr...  male  70.5      0      0      370369   7.7500   NaN        Q
170          171         0       1  Van der hoe...  male  61.0      0      0      111240  33.5000   B19        S

In [41]:
titanic_df[titanic_df['Age'] > 60][['Name','Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [42]:
titanic_df.loc[titanic_df['Age'] > 60, ['Name','Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [43]:
titanic_df[(titanic_df['Age'] > 60) & (titanic_df['Pclass'] == 1) & (titanic_df['Sex'] == 'female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

In [44]:
cond1 = titanic_df['Age'] > 60
cond2 = titanic_df['Pclass'] == 1
cond3 = titanic_df['Sex'] == 'female'
titanic_df[cond1 & cond2 & cond3]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

In [45]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [46]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

In [47]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [48]:
titanic_df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

In [52]:
titanic_groupby = titanic_df.groupby(by = 'Pclass')
print(type(titanic_groupby))
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

In [53]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [54]:
titanic_df.groupby('Pclass')['Age'].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [55]:
agg_format = {'Age' : 'max', 'SibSp' : 'sum', 'Fare' : 'mean'} # 원하는 형식으로 format 저장
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

In [56]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [57]:
titanic_df.isna().sum() # NaN값 확인

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [59]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head()

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S
3            4         1       1  Futrelle, M...  female  35.0      1      0          113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450   8.0500  C000        S

In [60]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [62]:
lambda_square = lambda x : x**2
print(lambda_square(3))

9


In [64]:
a = [1,2,3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

In [66]:
titanic_df['Name_len'] = titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

In [74]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

In [75]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<=15 else ('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

In [76]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성
def get_category(age):
    cat = ''
    if age <= 5: cat = 'Baby'
    elif age <= 12: cat = 'Child'
    elif age <= 18: cat = 'Teenager'
    elif age <= 25: cat = 'Student'
    elif age <= 35: cat = 'Young Adult'
    elif age <= 60: cat = 'Adult'
    else: cat = 'Elderly'
        
    return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult